src/main/java/com/denizenscript/clientizen/debuggui/DebugConsole.java

```diff
@@ -0,0 +1,145 @@
+			debugText = new StringBuilder();
+		});
+	}
+
+	public static ConsoleTextArea textArea;
+
+	public static void addDebug(String debug) {
+		if (!debugText.isEmpty()) {
+			debugText.append('\n').append(debug);
+		}
+		else {
+			debugText.append(debug);
+		}
+		if (textArea != null) {
+			textArea.addLine(debug);
+		}
+	}
+
+	public MinecraftClient client = MinecraftClient.getInstance();
+
+	public DebugConsole() {
+		super(new WPlainPanel());
+		Window window = client.getWindow();
+		setScrollingHorizontally(TriState.FALSE);
+		setBackgroundPainter(BackgroundPainter.createColorful(Color.BLACK.toRgb()));
+		setSize(window.getScaledWidth(), window.getScaledHeight() - ClientizenDebugGUI.TAB_HEIGHT);
+
+		children.remove(getVerticalScrollBar());
+		CustomScrollBar customScrollBar = new CustomScrollBar(Axis.VERTICAL, 20);
+		customScrollBar.setParent(this);
+		children.add(customScrollBar);
+		setVerticalScrollBar(customScrollBar);
+
+		getHeldPanel().setInsets(new Insets(5, 5, 0, 0));
+		textArea = new ConsoleTextArea();
+		getHeldPanel().setSize(window.getScaledWidth(), textArea.getHeight());
+		getHeldPanel().add(textArea, 0, 0);
+	}
+
+	public void onClose() {
+		textArea = null;
+	}
+
+	public WPlainPanel getHeldPanel() { // TODO: Mixins
+		return ReflectionHelper.getFieldValue(WScrollPanel.class, "widget", this);
+	}
+
+	public WScrollBar getVerticalScrollBar() { // TODO: Mixins
+		return ReflectionHelper.getFieldValue(WScrollPanel.class, "verticalScrollBar", this);
+	}
+
+	public void setVerticalScrollBar(WScrollBar scrollBar) { // TODO: Mixins
+		ReflectionHelper.setFieldValue(WScrollPanel.class, "verticalScrollBar", this, scrollBar);
+	}
+
+	public static class CustomScrollBar extends WScrollBar {
+
+		public int scrollSpeed;
+
+		public CustomScrollBar(Axis axis, int scrollSpeed) {
+			super(axis);
+			this.scrollSpeed = scrollSpeed;
+		}
+
+		@Override
+		public InputResult onMouseScroll(int x, int y, double amount) { // Same as super method, but uses a custom scrolling speed
+			setValue(getValue() + (int) -amount * scrollSpeed);
+			return InputResult.PROCESSED;
+		}
+	}
+
+	public class ConsoleTextArea extends WText {
+
+		public ConsoleTextArea() {
+			super(Text.literal(debugText.toString()));
+			setWrappedLines(new ArrayList<>(client.textRenderer.wrapLines(getText(), client.getWindow().getScaledWidth())));
+			updateSize(false);
+		}
+
+		public void addLine(String line) {
+			getWrappedLines().addAll(client.textRenderer.wrapLines(Text.literal(line), client.getWindow().getScaledWidth()));
+			updateSize(true);
+		}
+
+		public void updateSize(boolean updateScroll) {
+			setSize(width, client.textRenderer.fontHeight * getWrappedLines().size());
+			if (updateScroll) {
+				WScrollBar scrollBar = getVerticalScrollBar();
+				boolean wasAtBottom = scrollBar.getValue() == scrollBar.getMaxScrollValue();
+				DebugConsole.this.layout();
+				if (wasAtBottom) {
+					scrollBar.setValue(scrollBar.getMaxScrollValue());
+				}
```

Comment:
Perhaps put some inline comments here explaining what's going on since `layout` isn't enough of a description to determine that the scrollbar might change size

Reply:
Fair enough, already included the redundant `DebugConsole.this` to make it clearer what's being laid out but will add a comment as well